<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Подключение-к-базе-данных-и-исследование-таблиц" data-toc-modified-id="Подключение-к-базе-данных-и-исследование-таблиц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подключение к базе данных и исследование таблиц</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li></ul></li></ul></div>

# Исследование базы данных сервиса чтения книг по подписке

## Цель исследования

Cформулировать ценностное предложение для нового продукта.

## Описание данных

- **Таблица** `books`\
Содержит данные о книгах:\
`book_id` — идентификатор книги;\
`author_id` — идентификатор автора;\
`title` — название книги;\
`num_pages` — количество страниц;\
`publication_date` — дата публикации книги;\
`publisher_id` — идентификатор издателя.


- **Таблица** `authors`\
Содержит данные об авторах:\
`author_id` — идентификатор автора;\
`author` — имя автора.


- **Таблица** `publishers`\
Содержит данные об издательствах:\
`publisher_id` — идентификатор издательства;\
`publisher` — название издательства;


- **Таблица** `ratings`\
Содержит данные о пользовательских оценках книг:\
`rating_id` — идентификатор оценки;\
`book_id` — идентификатор книги;\
`username` — имя пользователя, оставившего оценку;\
`rating` — оценка книги.


- **Таблица** `reviews`\
Содержит данные о пользовательских обзорах на книги:\
`review_id` — идентификатор обзора;\
`book_id` — идентификатор книги;\
`username` — имя пользователя, написавшего обзор;\
`text`— текст обзора.

![map.png](map.png)

## Подключение к базе данных и исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
#------------------------------------------------------------------------------------------------------------------
# запрос books
#------------------------------------------------------------------------------------------------------------------
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
print(f"\033[1m{'BOOKS:'}")
display(pd.io.sql.read_sql(sql=text(query), con = con))
print('_' * 100)
#------------------------------------------------------------------------------------------------------------------
# запрос authors
#------------------------------------------------------------------------------------------------------------------
print(f"\033[1m{'AUTHORS:'}")
query = '''SELECT * FROM authors LIMIT 5'''
display(pd.io.sql.read_sql(sql=text(query), con = con))
print('_' * 100)
#------------------------------------------------------------------------------------------------------------------
# запрос ratings
#------------------------------------------------------------------------------------------------------------------
print(f"\033[1m{'RATINGS:'}")
query = '''SELECT * FROM ratings LIMIT 5'''
display(pd.io.sql.read_sql(sql=text(query), con = con))
print('_' * 100)
#------------------------------------------------------------------------------------------------------------------
# запрос reviews
#------------------------------------------------------------------------------------------------------------------
print(f"\033[1m{'REVIEWS:'}")
query = '''SELECT * FROM reviews LIMIT 5'''
display(pd.io.sql.read_sql(sql=text(query), con = con))
print('_' * 100)
#------------------------------------------------------------------------------------------------------------------
# запрос publishers
#------------------------------------------------------------------------------------------------------------------
print(f"\033[1m{'PUBLISHERS:'}")
query = '''SELECT * FROM publishers LIMIT 5'''
display(pd.io.sql.read_sql(sql=text(query), con = con))

BOOKS:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

____________________________________________________________________________________________________
AUTHORS:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

____________________________________________________________________________________________________
RATINGS:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

____________________________________________________________________________________________________
REVIEWS:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

____________________________________________________________________________________________________
PUBLISHERS:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [2]:
query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

<div style="border:dashed Orange 4px; padding: 10px; border-radius: 8px">
    <p style = 'color: white;background-color: blue; border-radius: 8px' >
        <b>Вывод:</b> </p>

Посчитав количество `book_id` из таблицы `book`, приминив фильтр более 2000-01-01 по полю `publication_date`, получили **819 книг** .
</div>

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [3]:
query = '''

SELECT books.book_id, 
       books.title, 
       COUNT(DISTINCT reviews.review_id) AS count_reviews, 
       ROUND(AVG(ratings.rating),1) AS avg_rating
FROM books

LEFT JOIN reviews
ON books.book_id = reviews.book_id

LEFT JOIN  ratings
ON books.book_id = ratings.book_id

GROUP BY books.book_id, books.title
ORDER BY count_reviews DESC

'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     count_reviews  avg_rating  
0                7         3.7  
1                6         4.0  
2                6         4.2  
3                6         4.4  
4                6         4.1  
..             ...         ...  
995              0         3.7  
996              0         3.0  
997              0         5.0  
998              0         4.0  
999              0         4.0  

[1000 rows x 4 columns]

<div style="border:dashed Orange 4px; padding: 10px; border-radius: 8px">
    <p style = 'color: white;background-color: blue; border-radius: 8px' >
        <b>Вывод:</b> </p>

Обьединив по `book_id` таблицы `book`, `reviews`, `ratings`, вывели на экран данные сгруппированые по `title` и отсортированные по убыванию `count_reviews`.
Получили данные на 1000 книг с количеством обзоров от 0 до 7 и средним рейтингом. Количество обзоров не влияет на рейтин книг, при 0 обзоров рейтинг книги может ровняться 5.
</div>

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [4]:
query = '''

WITH filter AS (SELECT publisher_id, book_id 
                FROM books 
                WHERE num_pages > 50)

SELECT filter.publisher_id, publishers.publisher
FROM filter 

INNER JOIN publishers
ON filter.publisher_id = publishers.publisher_id

GROUP BY filter.publisher_id, publishers.publisher
ORDER BY COUNT(filter.book_id) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher
0           212  Penguin Books

<div style="border:dashed Orange 4px; padding: 10px; border-radius: 8px">
    <p style = 'color: white;background-color: blue; border-radius: 8px' >
        <b>Вывод:</b> </p>

Определили фильтр по количеству страниц для книг и издательств из обьединенных таблиц `books`, `publishers`.
Сгруппированные и отсортированные по убыванию данные показывают издательство выпустившее наибольшее количество книг **Penguin Books**.

</div>

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [5]:
query = '''

WITH 

filter AS (SELECT book_id
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(rating_id) > 50)


SELECT authors.author_id, 
       authors.author, 
       AVG(ratings.rating) AS avg_rating
FROM books

INNER JOIN filter
ON books.book_id = filter.book_id

INNER JOIN authors
ON books.author_id = authors.author_id

INNER JOIN ratings
ON filter.book_id = ratings.book_id

GROUP BY authors.author_id, authors.author
ORDER BY AVG(ratings.rating) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

<div style="border:dashed Orange 4px; padding: 10px; border-radius: 8px">
    <p style = 'color: white;background-color: blue; border-radius: 8px' >
        <b>Вывод:</b> </p>

Определив фильт по оценкам книг таблицы `books`, выводим данные из обьединенной таблицы `books`, `authors`, `ratings`. Данные сгруппированы по автору, добавлена колонка с расчетом среднего рейтинга, отсортированного по убыванию. Автор с самой высокой средней оценкой книг - **J.K. Rowling**.

</div>

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [6]:
query = '''

SELECT COUNT(DISTINCT reviews.text)/ COUNT(DISTINCT reviews.username) AS avg_reviews_count
FROM ratings

INNER JOIN reviews
ON ratings.username = reviews.username

WHERE ratings.username in ( SELECT username
                            FROM ratings
                            GROUP BY username
                            HAVING COUNT(rating_id) > 48)

'''

pd.io.sql.read_sql(query, con = engine)

avg_reviews_count
0                 24

<div style="border:dashed Orange 4px; padding: 10px; border-radius: 8px">
    <p style = 'color: white;background-color: blue; border-radius: 8px' >
        <b>Вывод:</b> </p>

ИЗ обьединенной таблицы `raitings` и `reviews` получили данные о количестве обзоров на авторов обзоров.
Вывели данные с ограничением выше 48 оценок в `raiting_id` таблицы `raitings`. Среднее количество обзоров - **24**.

</div>